File: src/poe2_companion/lua/subjects.lua
local M={}
function M.resolve(build)
 local env=build.calcsTab.mainEnv
 local skill=env and env.player.mainSkill
 local active=skill and skill.activeEffect
 local effect=active and active.grantedEffect
 if not effect or not effect.id or not build.data.skills[effect.id] then return nil end
 local instance
 for setId,set in pairs(build.skillsTab.skillSets) do
  if set.socketGroupList==build.skillsTab.socketGroupList then
   for groupId,group in ipairs(set.socketGroupList) do
    for gemId,gem in ipairs(group.gemList or {}) do
     if gem==active.srcInstance then instance='skill:s'..setId..':g'..groupId..':n'..gemId end
    end
   end
  end
 end
 return {skill_instance_id=instance,skill_id=effect.id,actor_ref=skill.minion and 'minion' or 'player',
  component_ref=effect.id,weapon_set_id=build.itemsTab.activeItemSet.useSecondWeaponSet and 2 or 1}
end
function M.select(build,target,refresh)
 if not target then return nil end
 if target.actor_ref~='player' and target.actor_ref~='minion' then return 'actor_not_supported' end
 local s,g,n=target.skill_instance_id:match('^skill:s(%d+):g(%d+):n(%d+)$')
 s,g,n=tonumber(s),tonumber(g),tonumber(n)
 local set=build.skillsTab.skillSets[s]
 local group=set and set.socketGroupList[g]
 local gem=group and group.gemList[n]
 if not gem then return 'instance_not_found' end
 if group.enabled==false or gem.enabled==false then return 'instance_disabled' end
 build.skillsTab:SetActiveSkillSet(s,true)
 build.itemsTab.controls['weaponSwap'..target.weapon_set_id]:Click()
 build.mainSocketGroup=g
 refresh()
 local matches={}
 for index,skill in ipairs(group.displaySkillList or {}) do
  local active=skill.activeEffect
  if active and active.srcInstance==gem and
   (not target.component_ref or active.grantedEffect.id==target.component_ref) then
   matches[#matches+1]=index
  end
 end
 if #matches==0 then return 'component_not_found' end
 if #matches>1 then return 'ambiguous_component' end
 group.mainActiveSkill=matches[1]
 refresh()
 return nil
end
function M.bind(build,target,saved,selectionError,scenarioDigest)
 local evaluated=M.resolve(build)
 local reason=selectionError
 if target and not reason then
  local main=build.calcsTab.mainEnv.player.mainSkill
  if not evaluated or evaluated.skill_instance_id~=target.skill_instance_id or
   evaluated.weapon_set_id~=target.weapon_set_id or main.disableReason or
   (target.component_ref and evaluated.component_ref~=target.component_ref) then
   reason='eligibility_changed'
  elseif evaluated.actor_ref~=target.actor_ref then reason='actor_mismatch' end
 end
 local nextAction=reason and 'inspect_skill_instances' or nil
 if reason=='ambiguous_component' then nextAction='choose_component'
 elseif reason=='actor_not_supported' or reason=='actor_mismatch' then nextAction='supply_supported_actor'
 elseif reason=='eligibility_changed' or reason=='instance_disabled' then nextAction='repair_skill_eligibility' end
 return {saved=saved,requested=target,evaluated=not reason and evaluated or nil,
  status=reason and 'unavailable' or target and 'matched' or 'saved_default',
  reason=reason,scenario_digest=scenarioDigest,next_action=nextAction}
end
return M

File: src/poe2_companion/subjects.py
"""Explicit, discovered calculation identities shared across every candidate."""
from typing import Annotated, Literal, Self
from pydantic import Field, model_validator
from .builds import DTO
from .profiles import SkillInstanceID, Digest

EffectID = Annotated[str, Field(pattern=r'^[A-Za-z0-9_]+$', max_length=120)]
Actor = Literal['player', 'minion', 'hollow_image', 'spirit_vessel']


class CalculationTarget(DTO):
    skill_instance_id: SkillInstanceID
    actor_ref: Actor
    component_ref: EffectID | None = None
    weapon_set_id: Literal[1, 2]
    aggregation: Literal['single_skill', 'component_breakdown'] = 'single_skill'


class EvaluatedSubject(DTO):
    skill_instance_id: SkillInstanceID | None = None
    skill_id: EffectID
    actor_ref: Actor
    component_ref: EffectID
    weapon_set_id: Literal[1, 2]


class SubjectBinding(DTO):
    saved: EvaluatedSubject | None = None
    requested: CalculationTarget | None = None
    evaluated: EvaluatedSubject | None = None
    status: Literal['saved_default', 'matched', 'unavailable']
    reason: Literal['instance_not_found', 'instance_disabled', 'ambiguous_component',
        'component_not_found', 'actor_not_supported', 'actor_mismatch', 'eligibility_changed'] | None = None
    scenario_digest: Digest
    next_action: Literal['inspect_skill_instances', 'choose_component', 'supply_supported_actor', 'repair_skill_eligibility'] | None = None

    @model_validator(mode='after')
    def coherent(self) -> Self:
        if self.status=='matched':
            if self.requested is None or self.evaluated is None:
                raise ValueError('matched_subject_requires_identities')
            if (self.requested.skill_instance_id!=self.evaluated.skill_instance_id
                    or self.requested.actor_ref!=self.evaluated.actor_ref
                    or self.requested.weapon_set_id!=self.evaluated.weapon_set_id
                    or (self.requested.component_ref is not None and self.requested.component_ref!=self.evaluated.component_ref)):
                raise ValueError('evaluated_subject_mismatch')
        if self.status=='unavailable' and self.evaluated is not None:
            raise ValueError('unavailable_subject_cannot_be_evaluated')
        if self.status=='saved_default' and self.requested is not None:
            raise ValueError('explicit_target_cannot_fall_back_to_saved_default')
        return self
